Add sensor platform for SolarEdge Monitoring API

GidoHakvoort opened the pull request:
## Description:

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7735

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: solaredge
    api_key: YOUR_API_KEY
    site_id: YOUR_SITE_ID
    name: SolarEdge
    monitored_conditions:
      - currentPower
      - lastDayData
      - lastMonthData
      - lastYearData

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

GidoHakvoort committed:
Adding sensor for SolarEdge Monitoring API support

GidoHakvoort committed:
Adding support for Rova garbage calendar

GidoHakvoort committed:
Update solaredge to pass lint and flake8

GidoHakvoort committed:
Added solaredge.py to .coveragerc

GidoHakvoort committed:
Added extend for Voluptuous schema

GidoHakvoort committed:
Fixed styling issues

GidoHakvoort committed:
Removed rova.py for later feature

homeassistant commented:
Hello @GidoHakvoort,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`df98c0a73abf4618e921a6c84296e274e596147b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;18846&#x2F;commits&#x2F;df98c0a73abf4618e921a6c84296e274e596147b) - This commit has something that looks like an email address (gido@hakvoort.it). Maybe try linking that to GitHub?.
- [`12e3d43cb3c91d2843ebe8231c64e9922ffebbe1`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;18846&#x2F;commits&#x2F;12e3d43cb3c91d2843ebe8231c64e9922ffebbe1) - This commit has something that looks like an email address (gido@hakvoort.it). Maybe try linking that to GitHub?.
- [`b74cde316590f9528ae66474ad7dbfc6e14c21fd`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;18846&#x2F;commits&#x2F;b74cde316590f9528ae66474ad7dbfc6e14c21fd) - This commit has something that looks like an email address (gido@hakvoort.it). Maybe try linking that to GitHub?.
- [`4033255dd57de15728df6ad1f234cf5f32dd741c`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;18846&#x2F;commits&#x2F;4033255dd57de15728df6ad1f234cf5f32dd741c) - This commit has something that looks like an email address (gido@hakvoort.it). Maybe try linking that to GitHub?.
- [`3b4b13d1d492aa0aa6c7f1128166739bc8638c2e`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;18846&#x2F;commits&#x2F;3b4b13d1d492aa0aa6c7f1128166739bc8638c2e) - This commit has something that looks like an email address (gido@hakvoort.it). Maybe try linking that to GitHub?.
- [`85f9f36976e47fe366b72a0cfda9122952b48a6e`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;18846&#x2F;commits&#x2F;85f9f36976e47fe366b72a0cfda9122952b48a6e) - This commit has something that looks like an email address (gido@hakvoort.it). Maybe try linking that to GitHub?.
- [`63d07d1b19ad953e3d2cbf30166611638b0c1ebc`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;18846&#x2F;commits&#x2F;63d07d1b19ad953e3d2cbf30166611638b0c1ebc) - This commit has something that looks like an email address (gido@hakvoort.it). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

pvizeli reviewed (changes requested):
https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

Please create a python pip packages for you API

GidoHakvoort committed:
Replaced API requests with python pip package

GidoHakvoort committed:
Fixed styling issues

GidoHakvoort commented:
@pvizeli Thanks, should be fixed

fabaff commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,161 @@
+                         discovery_info=None):
+    """Create the SolarEdge Monitoring API sensor."""
+    api_key = config.get(CONF_API_KEY, None)
+    site_id = config.get(CONF_SITE_ID, None)
```
`None` is not needed it's the default while using `dict.get()` and both are required and validated already. Thus we prefer it `dict[key]` is used.

fabaff commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,161 @@
+class SolarEdgeSensor(Entity):
+    """Representation of an SolarEdge Monitoring API sensor."""
+
+    def __init__(self, sensorType, data):
```
"snake case" please.

fabaff commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,161 @@
+    @property
+    def name(self):
+        """Return the name."""
+        return 'solaredge_' + self.type
```
Would be nice if the user could overwrite this directly with the platform configuration -> `CONF_NAME` with a default of "SolarEdge".

fabaff commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,161 @@
+        from requests.exceptions import HTTPError, ConnectTimeout
+
+        # Create new SolarEdge object to retrieve data
+        api = solaredge.Solaredge(self.api_key)
```
This should be setup in `async_setup_platform()` and passed down. During `async_setup_platform()` check if the credentials are valid. If not, make the setup fail. This way the users will not end-up with a non-working platform.

fabaff commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,161 @@
+        self.site_id = site_id
+
+    @asyncio.coroutine
+    def schedule_update(self, minutes):
```
Remove this and add a throttle to `async_update()` and perhaps introduce `available()` if the API is unstable.

GidoHakvoort committed:
Updated to new async syntax
Added async_update to sensor class
Added Throttle to SolarEdge data update
Added CONF_NAME to platform settings
Added credentials check for api
Minor code style changes

GidoHakvoort committed:
Remove unnecessary debug logging

Mariusthvdb commented:
HI, 

great to see this is being developed. Please allow me to ask while you're still at it, if you will also be adding the Inventory sensor, next to the planned Overview sensors:
fyi, this is how these are/(were with json) extracted, maybe you can add them, since they require the same site id and key:

```
- platform: jsonrest
  resource: !secret se_inventory_resource
  method: GET
  name: SolarEdge Inventory
  scan_interval: 3600

- platform: jsonrest
  resource: !secret se_overview_resource
  method: GET
  name: SolarEdge Overview
  scan_interval: 3600

- platform: template
  sensors:
    solaredge_overview_last_updated:
      friendly_name: Last updated
      value_template: "{{ states.sensor.solaredge_overview.attributes.overview.lastUpdateTime | timestamp_custom ('%H:%M') }}"
    solaredge_overview_lifetime:
      friendly_name: Lifetime
      value_template: '{{ (states.sensor.solaredge_overview.attributes.overview.lifeTimeData.energy | float / 1000000) | round(2) }}'
      unit_of_measurement: 'MWh'
    solaredge_overview_last_year:
      friendly_name: Last year
      value_template: '{{ (states.sensor.solaredge_overview.attributes.overview.lastYearData.energy | float / 1000) | round(1) }}'
      unit_of_measurement: 'kWh'
    solaredge_overview_last_month:
      friendly_name: Last month
      value_template: '{{ (states.sensor.solaredge_overview.attributes.overview.lastMonthData.energy | float / 1000) | round(1) }}'
      unit_of_measurement: 'kWh'
    solaredge_overview_last_day:
      friendly_name: Last day
      value_template: '{{ (states.sensor.solaredge_overview.attributes.overview.lastDayData.energy | float / 1000) | round(1) }}'
      unit_of_measurement: 'kWh'
    solaredge_overview_current_power:
      friendly_name: Current power
      value_template: '{{ (states.sensor.solaredge_overview.attributes.overview.currentPower.power | float / 1000) | round(2) }}'
      unit_of_measurement: 'kW'

    solaredge_inventory_name:
      friendly_name: Name
      value_template: '{{states.sensor.solaredge_inventory.attributes.Inventory.inverters[0].name}}'
    solaredge_inventory_manufacturer:
      friendly_name: Mnf
      value_template: '{{states.sensor.solaredge_inventory.attributes.Inventory.inverters[0].manufacturer}}'
    solaredge_inventory_model:
      friendly_name: Model
      value_template: '{{states.sensor.solaredge_inventory.attributes.Inventory.inverters[0].model}}'
    solaredge_inventory_communication:
      friendly_name: Comm
      value_template: '{{states.sensor.solaredge_inventory.attributes.Inventory.inverters[0].communicationMethod}}'
    solaredge_inventory_dsp1:
      friendly_name: Dsp1
      value_template: '{{states.sensor.solaredge_inventory.attributes.Inventory.inverters[0].dsp1Version}}'
    solaredge_inventory_dsp2:
      friendly_name: Dsp2
      value_template: '{{states.sensor.solaredge_inventory.attributes.Inventory.inverters[0].dsp2Version}}'
    solaredge_inventory_cpu:
      friendly_name: Cpu
      value_template: '{{states.sensor.solaredge_inventory.attributes.Inventory.inverters[0].cpuVersion}}'
    solaredge_inventory_sn:
      friendly_name: Sn
      value_template: '{{states.sensor.solaredge_inventory.attributes.Inventory.inverters[0].SN}}'
    solaredge_inventory_optimizers:
      friendly_name: Opt
      value_template: '{{states.sensor.solaredge_inventory.attributes.Inventory.inverters[0].connectedOptimizers}}'
```

maybe even consider site-details and equipment-list?
thanks.

GidoHakvoort commented:
Hi @Mariusthvdb,

Thanks, nice suggestion, I will look into it. For the site details I could add an additional sensor which will be updated once a day (SolarEdge has a daily query limit of 300 requests). Details such as site id, name, status, peak power, etc. could then be added as attributes to a sensor (e.g. solaredge_site_details). For the inventory, I might be able to separate this per equipment type (e.g. meters, sensors, gateways, etc). This might however be a bit more difficult as these are lists. If you have any suggestions, let me know.

GidoHakvoort committed:
Updated dict keys

pvizeli commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,175 @@
+# Config for solaredge monitoring api requests.
+CONF_SITE_ID = "site_id"
+
+UPDATE_DELAY = timedelta(minutes=10)
```
Add also a `SCAN_INTERVAL` with 10 minutes

pvizeli commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,175 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, async_add_entities,
```
Your platform is not async. Remove this prefix. Otherwise, you need to use aiohttp.

GidoHakvoort committed:
Added SCAN_INTERVAL
Updated platform setup

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+
+REQUIREMENTS = ['solaredge==0.0.2']
+
+DOMAIN = "solaredge"
```
Please remove this constant. This platform belongs to the `'sensor'` domain.

GidoHakvoort replied:
Hi @MartinHjelmare thanks for the review, all issues should be fixed.

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+
+import voluptuous as vol
+
+from homeassistant.components.light import PLATFORM_SCHEMA
```
Don't import `PLATFORM_SCHEMA` from the light component. Import it from the sensor component.

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Required(CONF_SITE_ID): cv.string,
+    vol.Optional(CONF_NAME, default='SolarEdge'): cv.string,
+    vol.Optional(CONF_MONITORED_VARIABLES, default=['current_power']):
```
We prefer to use `CONF_MONITORED_CONDITIONS` for new integrations. Import that from const.py.

GidoHakvoort replied:
Changed this to CONF_MONITORED_CONDITIONS.

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+    site_id = config[CONF_SITE_ID]
+    platform_name = config[CONF_NAME]
+
+    _LOGGER.debug("Setting up SolarEdge Monitoring API")
```
We can remove this. It's already logged by the core.

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+
+        if response['details']['status'].lower() != 'active':
+            _LOGGER.debug("SolarEdge site is not active")
+            return False
```
Nothing is checking this return value. Just `return`.
```suggestion
            return
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+        _LOGGER.debug("Credentials correct and site is active")
+    except KeyError:
+        _LOGGER.debug("Missing details data in solaredge response")
+        return False
```
```suggestion
        return
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+            return False
+        _LOGGER.debug("Credentials correct and site is active")
+    except KeyError:
+        _LOGGER.debug("Missing details data in solaredge response")
```
```suggestion
        _LOGGER.error("Missing details data in solaredge response")
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+        response = api.get_details(site_id)
+
+        if response['details']['status'].lower() != 'active':
+            _LOGGER.debug("SolarEdge site is not active")
```
```suggestion
            _LOGGER.error("SolarEdge site is not active")
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+        _LOGGER.debug("Missing details data in solaredge response")
+        return False
+    except (ConnectTimeout, HTTPError):
+        _LOGGER.debug("Could not retrieve details from SolarEdge \
```
Don't break the line with backslash. Break it directly after the parenthesis, and if we need more lines for the logging string, just end the string with a quote, and continue on next line with a quote.

Also change log level to error.

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+    except (ConnectTimeout, HTTPError):
+        _LOGGER.debug("Could not retrieve details from SolarEdge \
+         Monitoring API")
+        return False
```
```suggestion
        return
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+        self._state = None
+
+        self._json_key = SENSOR_TYPES[self.sensor_key][0]
+        self._name = SENSOR_TYPES[self.sensor_key][1]
```
This isn't used.

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+        """Return the state of the sensor."""
+        return self._state
+
+    async def async_update(self):
```
```suggestion
    def update(self):
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+
+    async def async_update(self):
+        """Get the latest data from the sensor and update the state."""
+        _LOGGER.debug("async_update")
```
Remove this.

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+    async def async_update(self):
+        """Get the latest data from the sensor and update the state."""
+        _LOGGER.debug("async_update")
+        await self.hass.async_add_job(self.data.update)
```
```suggestion
        self.data.update()
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+        self.data = {}
+        self.site_id = site_id
+
+        self.update()
```
We can remove this. Both entity update and data update will be called during entity addition since we call `add_entities` with second argument `True`.

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+            data = self.api.get_overview(self.site_id)
+            overview = data['overview']
+        except KeyError:
+            _LOGGER.debug("Missing overview data, skipping update")
```
```suggestion
            _LOGGER.error("Missing overview data, skipping update")
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+            _LOGGER.debug("Missing overview data, skipping update")
+            return
+        except (ConnectTimeout, HTTPError):
+            _LOGGER.debug("Could not retrieve data, skipping update")
```
```suggestion
            _LOGGER.error("Could not retrieve data, skipping update")
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+
+        self.data = {}
+
+        for item in overview:
```
If `overview` is a dict we can do:
```suggestion
        for key, value in overview.items():
```

MartinHjelmare commented on homeassistant/components/sensor/solaredge.py:
```diff
@@ -0,0 +1,176 @@
+                self.data[item] = value['power']
+
+        _LOGGER.debug("Updated SolarEdge overview data: %s", self.data)
+        return
```
Not needed return.

GidoHakvoort committed:
Remove DOMAIN variable
Correct import for PLATFORM_SCHEMA

GidoHakvoort committed:
Change some debug to error messages
Correct return statements
Remove initial update call

GidoHakvoort committed:
Fix pylint and flake8 errors

MartinHjelmare reviewed (approved):
Looks good! Can be merged when build passes.
